dataactvalidator/scripts/load_program_activity.py

```diff
@@ -20,54 +29,179 @@ def get_program_activity_file(base_path):
+            s3bucket connection object
+    """
+    if CONFIG_BROKER["use_aws"]:
+        s3 = boto3.resource('s3')
+        s3_object = s3.Object(PA_BUCKET, PA_SUB_KEY + PA_FILE_NAME)
+        return s3_object
+    raise Exception("Config specifies not to use AWS, cannot call S3 bucket.")
+
+
+def get_date_of_current_pa_upload(base_path):
+    """ Gets the last time the file was uploaded to S3, or alternatively the last time the local
+        file was modified.
+
+        Args:
+            base_path: directory of domain config files
+
+        Returns:
+            DateTime object
+    """
+    if CONFIG_BROKER["use_aws"]:
+        last_uploaded = boto3.client('s3').head_object(Bucket=PA_BUCKET, Key=PA_SUB_KEY+PA_FILE_NAME)['LastModified']
+    else:
+        pa_file = get_program_activity_file(base_path)
+        last_uploaded = datetime.datetime.utcfromtimestamp(os.path.getmtime(pa_file))
+    return last_uploaded
+
+
+def get_stored_pa_last_upload():
+    """ Gets last recorded timestamp from last time file was processed.
+
+        Returns:
+            Upload date of most recent file we have recorded (Datetime object)
+    """
+    sess = GlobalDB.db().session
+    last_stored_obj = sess.query(ExternalDataLoadDate).join(ExternalDataType).filter(
+        ExternalDataType.name == "program_activity_upload"
+        ).one_or_none()
+    if not last_stored_obj:
+        # return epoch ts to make sure we load the data the first time through,
+        # and ideally any time the data might have been wiped
+        last_stored = datetime.datetime.utcfromtimestamp(0)
+    else:
+        last_stored = last_stored_obj.last_load_date
+    return last_stored
+
+
+def set_stored_pa_last_upload(load_datetime):
+    """ Set upload date of most recent file we have recorded (Datetime object)
+
+        Args:
+            Datetime object representing the timestamp associated with the current file
+    """
+    sess = GlobalDB.db().session
+    last_stored_obj = sess.query(ExternalDataLoadDate).join(ExternalDataType).filter(
+        ExternalDataType.name == "program_activity_upload"
+        ).one_or_none()
+    if not last_stored_obj:
+        data_type = sess.query(ExternalDataType).filter(
+            ExternalDataType.name == "program_activity_upload"
+            ).one_or_none()
+        last_stored_obj = ExternalDataLoadDate(external_data_type_id=data_type.external_data_type_id,
+                                               last_load_date=load_datetime)
+        sess.add(last_stored_obj)
+    else:
+        last_stored_obj.last_load_date = load_datetime
+    sess.commit()
+
+
 def load_program_activity_data(base_path):
     """ Load program activity lookup table.
 
         Args:
             base_path: directory of domain config files
     """
+    last_upload = get_date_of_current_pa_upload(base_path)
+    if not (last_upload > get_stored_pa_last_upload()):
+        return
+    else:
+        set_stored_pa_last_upload(last_upload)
```

Comment:
A) Do we need an `else` here since the return would get us out of this function entirely?
B) Do we want to set it now when the process might still fail or do we want to set it at the end when the process is successfully completed?

Reply:
we should be setting this at the end, only when it succeeds. Will refactor this.